### PR TITLE
[PATCH v4] validation: dma: verify maximum concurrent transfers

### PR DESCRIPTION
**Pull request by TuomasTaipale**
Add and modify tests to verify that maximum signaled in-flight transfer count is handled correctly.

With synchronous transfers, run more than maximum in-flight transfer count back to back to verify proper implementation resource handling.

With asynchronous transfers, start maximum in-flight count of transfers and wait for completion (poll/event).

v2:
- Added testing of maximum completion event allocation
- Added a new test for max DMA pools
- Took maximum segment length into account when configuring transfers

v3:
- Matias' comments

v4:
- Added reviewed-by tag

### CODE REVIEW

File: test/validation/api/dma/dma.c [hunk 1]
```diff
@@ -1000,7 +1063,7 @@ static void test_dma_addr_to_addr_sync(void)
 
 static void test_dma_addr_to_addr_sync_mtrs(void)
 {
-	test_dma_addr_to_addr_trs(ODP_DMA_COMPL_SYNC, 2, 0, 0);
+	test_dma_addr_to_addr_trs(ODP_DMA_COMPL_SYNC, global.dma_capa.max_transfers * 2, 0, 0);
```
Comment (MatiasElo) [inline test/validation/api/dma/dma.c:1066]: Perhaps it would be good to add a comment here to explain why a larger transfer count is used.

File: test/validation/api/dma/dma.c [hunk 2]
```diff
@@ -636,6 +641,64 @@ static void test_dma_addr_to_addr_trs(odp_dma_compl_mode_t compl_mode_mask, uint
...
+	odp_dma_transfer_param_t trs_param[num_trs];
+	odp_dma_t dma;
+	odp_dma_seg_t src_seg[num_trs];
+	odp_dma_seg_t dst_seg[num_trs];
+	int ret;
+	uint32_t i, cur_len;
+	uint8_t *src = global.src_addr + OFFSET;
+	uint8_t *dst = global.dst_addr + OFFSET;
+	uint32_t len = global.len;
+	uint32_t seg_len = len / num_trs;
```
Comment (MatiasElo) [inline test/validation/api/dma/dma.c:657]: Depending on `max_transfers` capa this could theoretically round down to zero.

File: test/validation/api/dma/dma.c [hunk 3]
```diff
@@ -380,126 +379,132 @@ static int do_transfer(odp_dma_t dma, const odp_dma_transfer_param_t *trs_param,
...
-		if (multi) {
-			const odp_dma_compl_param_t *compl_ptr[1] = {&compl_param};
-			const odp_dma_transfer_param_t *trs_ptr[1] = {trs_param};
-
-			ret = odp_dma_transfer_start_multi(dma, trs_ptr, compl_ptr, 1);
-		} else {
-			ret = odp_dma_transfer_start(dma, trs_param, &compl_param);
-		}
+		if (multi)
+			ret = odp_dma_transfer_start_multi(dma, trs_ptr, compl_ptr, num_trs);
```
Comment (MatiasElo) [inline test/validation/api/dma/dma.c:432]: Depending on the implementation, if `num_trs` is large all transfers may not be started with a single `odp_dma_transfer_start_multi()` call. Perhaps it would be better to check that the return value is  > 0 and repeat call enough times.

File: test/validation/api/dma/dma.c (hunk 2 above)
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c:657]: Yes, could be safer to incorporate `max_transfers` to the SHM calculation. The size will skyrocket though if the implementation is a capable one.

File: test/validation/api/dma/dma.c [hunk 4]
```diff
@@ -282,7 +281,7 @@ static void test_dma_compl_pool(void)
 	CU_ASSERT(strcmp(pool_info.name, name) == 0);
 	CU_ASSERT(pool_info.pool_ext == 0);
 	CU_ASSERT(pool_info.type == ODP_POOL_DMA_COMPL);
-	CU_ASSERT(pool_info.dma_pool_param.num == NUM_COMPL);
+	CU_ASSERT(pool_info.dma_pool_param.num == global.dma_capa.max_transfers);
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c:284]: Yes, good idea, I'll add the new test case as well.

[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c:289]: New patches for these in v2 (75ac960599546cb869557b6b9c7b9e7bdf3be225 and 8c9613dde17bf7e056dbefc87f13c0a77f8e763b).

File: test/validation/api/dma/dma.c (hunk 2 above)
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c]: Added a new patch for the max segment length stuff in v2, hopefully now properly taken into account (3069d430c8a50d783b5a52f88fc84680aef3fd7a).

File: test/validation/api/dma/dma.c [hunk 5]
```diff
@@ -329,6 +329,29 @@ static void test_dma_compl_pool_same_name(void)
...
+	odp_dma_pool_param_init(&dma_pool_param);
+	dma_pool_param.num = global.dma_capa.max_transfers;
+
+	for (i = 0; i < num; i++) {
+		pools[i] = odp_dma_pool_create(NULL, &dma_pool_param);
+
+		if (pools[i] == ODP_POOL_INVALID)
+			ODPH_ERR("DMA completion pool create failed: %u / %u\n", i, num);
+
+		CU_ASSERT_FATAL(pools[i] != ODP_POOL_INVALID);
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c]: V3, similar treatment for destruction phase.

File: test/validation/api/dma/dma.c [hunk 6]
```diff
@@ -329,6 +329,29 @@ static void test_dma_compl_pool_same_name(void)
 	CU_ASSERT_FATAL(odp_pool_destroy(pool_b) == 0);
 }
 
+static void test_dma_compl_pool_max_pools(void)
+{
+	odp_dma_pool_param_t dma_pool_param;
+	/* Max pools minus the ones already created in global init */
+	uint32_t num = global.dma_capa.pool.max_pools - 2, i, j;
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c:337]: Yes, there's a disclaimer about this in commit message but spec clarification would be more helpful.

File: test/validation/api/dma/dma.c [hunk 7]
```diff
@@ -288,14 +289,20 @@ static void test_dma_compl_pool(void)
...
+		u64 = odp_dma_compl_to_u64(compl[i]);
+		CU_ASSERT(u64 != odp_dma_compl_to_u64(ODP_DMA_COMPL_INVALID));
+		printf("\n    DMA compl handle: 0x%" PRIx64 "\n", u64);
+		odp_dma_compl_print(compl[i]);
 
-	u64 = odp_dma_compl_to_u64(compl);
-	CU_ASSERT(u64 != odp_dma_compl_to_u64(ODP_DMA_COMPL_INVALID));
-	printf("\n    DMA compl handle: 0x%" PRIx64 "\n", u64);
-	odp_dma_compl_print(compl);
+		if (compl[i] == ODP_DMA_COMPL_INVALID)
```
[the remark this answers is not in the document]
Comment (TuomasTaipale) [reply test/validation/api/dma/dma.c:298]: V3.

**Review by MatiasElo — APPROVED** [review]
